tree-walk/src/interpreter.cpp

```diff
@@ -279,6 +379,27 @@ Value InterpreterVisitor::operator()(const Literal &ltrl) {
   }
 }
 
+Value InterpreterVisitor::operator()(const Set &set) {
+  Value obj = std::visit(*this, *set.object);
+  if (!std::holds_alternative<ClsInstShrdPtr>(obj)) {
+    throw InterpretException("Tried to set a property on non class instance " +
+                             std::visit(s_valuePrinter, obj));
+  }
+
+  Value val = std::visit(*this, *set.value);
+  // Create a copy of the function and rename it
+  if (std::holds_alternative<FnClosureShrdPtr>(val)) {
+    FnClosureShrdPtr fnCopy =
+        std::make_shared<FunctionClosure>(*std::get<FnClosureShrdPtr>(val));
+    fnCopy->setName(set.property);
+    fnCopy->setIsInitialiser(set.property == "init");
+    val = fnCopy;
+  }
+  std::get<ClsInstShrdPtr>(obj)->getClosure()->upsert(std::string(set.property),
```

Comment:
Ya, this should be `upsertInScope` if we're going to keep this logic in Environment. It probably does belong in Environment.